File: frontend/src/api/groups.ts
import apiClient from './client'

export interface GroupItem {
  id: number
  name: string
  sort_order: number
  created_at: string
  updated_at: string
}

interface GroupListResponse {
  code: string
  data: GroupItem[]
}

interface GroupResponse {
  code: string
  data: GroupItem
}

export interface GroupPayload {
  name: string
  sort_order?: number
}

export async function listNodeGroups() {
  const { data } = await apiClient.get<GroupListResponse>('/api/v1/node-groups')
  return data
}

export async function createNodeGroup(payload: GroupPayload) {
  const { data } = await apiClient.post<GroupResponse>('/api/v1/node-groups', payload)
  return data
}

export async function updateNodeGroup(id: number, payload: GroupPayload) {
  const { data } = await apiClient.put<GroupResponse>(`/api/v1/node-groups/${id}`, payload)
  return data
}

export async function deleteNodeGroup(id: number) {
  const { data } = await apiClient.delete<{ code: string; message: string }>(`/api/v1/node-groups/${id}`)
  return data
}

export async function listSubscriptionGroups() {
  const { data } = await apiClient.get<GroupListResponse>('/api/v1/subscription-groups')
  return data
}

export async function createSubscriptionGroup(payload: GroupPayload) {
  const { data } = await apiClient.post<GroupResponse>('/api/v1/subscription-groups', payload)
  return data
}

export async function updateSubscriptionGroup(id: number, payload: GroupPayload) {
  const { data } = await apiClient.put<GroupResponse>(`/api/v1/subscription-groups/${id}`, payload)
  return data
}

export async function deleteSubscriptionGroup(id: number) {
  const { data } = await apiClient.delete<{ code: string; message: string }>(
    `/api/v1/subscription-groups/${id}`,
  )
  return data
}

File: backend/src/api/groups.rs
use axum::{
    Json,
    extract::{Path, State},
    http::HeaderMap,
};

use crate::{
    dto::groups::{CreateGroupRequest, GroupListResponse, GroupResponse, UpdateGroupRequest},
    errors::AppError,
    repository::group_repo::GroupTable,
    services::group_service,
    state::AppState,
};

pub async fn list_node_groups(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> Result<Json<GroupListResponse>, AppError> {
    group_service::require_auth(&state, &headers).await?;
    Ok(Json(
        group_service::list_groups(&state, GroupTable::Node).await?,
    ))
}

pub async fn create_node_group(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(payload): Json<CreateGroupRequest>,
) -> Result<Json<GroupResponse>, AppError> {
    group_service::require_auth(&state, &headers).await?;
    Ok(Json(
        group_service::create_group(&state, GroupTable::Node, payload).await?,
    ))
}

pub async fn update_node_group(
    State(state): State<AppState>,
    headers: HeaderMap,
    Path(id): Path<i64>,
    Json(payload): Json<UpdateGroupRequest>,
) -> Result<Json<GroupResponse>, AppError> {
    group_service::require_auth(&state, &headers).await?;
    Ok(Json(
        group_service::update_group(&state, GroupTable::Node, id, payload).await?,
    ))
}

pub async fn delete_node_group(
    State(state): State<AppState>,
    headers: HeaderMap,
    Path(id): Path<i64>,
) -> Result<Json<serde_json::Value>, AppError> {
    group_service::require_auth(&state, &headers).await?;
    group_service::delete_group(&state, GroupTable::Node, id).await?;
    Ok(Json(
        serde_json::json!({ "code": "00000", "message": "group deleted" }),
    ))
}

pub async fn list_subscription_groups(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> Result<Json<GroupListResponse>, AppError> {
    group_service::require_auth(&state, &headers).await?;
    Ok(Json(
        group_service::list_groups(&state, GroupTable::Subscription).await?,
    ))
}

pub async fn create_subscription_group(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(payload): Json<CreateGroupRequest>,
) -> Result<Json<GroupResponse>, AppError> {
    group_service::require_auth(&state, &headers).await?;
    Ok(Json(
        group_service::create_group(&state, GroupTable::Subscription, payload).await?,
    ))
}

pub async fn update_subscription_group(
    State(state): State<AppState>,
    headers: HeaderMap,
    Path(id): Path<i64>,
    Json(payload): Json<UpdateGroupRequest>,
) -> Result<Json<GroupResponse>, AppError> {
    group_service::require_auth(&state, &headers).await?;
    Ok(Json(
        group_service::update_group(&state, GroupTable::Subscription, id, payload).await?,
    ))
}

pub async fn delete_subscription_group(
    State(state): State<AppState>,
    headers: HeaderMap,
    Path(id): Path<i64>,
) -> Result<Json<serde_json::Value>, AppError> {
    group_service::require_auth(&state, &headers).await?;
    group_service::delete_group(&state, GroupTable::Subscription, id).await?;
    Ok(Json(
        serde_json::json!({ "code": "00000", "message": "group deleted" }),
    ))
}
